tsl/src/continuous_aggs/materialize.c

```diff
@@ -197,8 +198,21 @@ continuous_agg_materialize(int32 materialization_id, ContinuousAggMatOptions *op
 
 	catalog = ts_catalog_get();
 
-	/* copy over all the materializations from the raw hypertable to all the continuous aggs */
+	/* copy over all the materializations from the raw hypertable to all the continuous aggs
+	 */
 	caggs = ts_continuous_aggs_find_by_raw_table_id(cagg_data.raw_hypertable_id);
+	if (list_length(caggs) > 1)
```

Comment:
Do I understand correctly that the lock is only obtained when there are more than one cagg is defined, so there is a risk of concurrency? What's about a corner case when second cagg is created after materialising the first one passed this line? Can it be a race condition if meterializing the first one is slow or interrupted?
Or is this condition to deal with a deadlock?

Reply:
@k-rus @erimatnor Yes, the corner case is possible. But not fully convinced that we should deal with the error for the corner case. The locks have a concurrency cost. So acquiring a lock only if it is necessary to do so seems to be a good trade off to me. ( There is an additional optimization possible here : a first time cagg does not have to deal with any invalidations since it has to fully materialize all the data. But we don't optimize for that case. Then this code might actually not be exercised in this scenario).

Reply:
Do we know how big this cost is and do we throw a reasonable error when we hit the corner case? If not, I think we should cover the corner-case too and care about "correctness" prior to optimizing. We can always optimize further at a later time.

Reply:
I would like to add: which PG processes will be affected by this cost? Will insertion to hypertable or querying hypertable be affected by this concurrency overhead?